005-windows-handles-not-fd.md

```diff
@@ -0,0 +1,62 @@
+| Title  | Windows HANDLEs not fds |
+|--------|-------------------------|
+| Author | @vtjnash                |
+| Status | DRAFT                   |
+| Date   | 2016-09-27              |
+
+
+## Overview
+
+Currently the libuv API requires passing in a MSVCRT-emulated file-descriptor,
+and then immediately discards the wrapper and uses the native HANDLE instead.
+This LEP proposes is to consistently remove this extra indirection from all APIs in libuv,
+and replace all usages of `uv_os_sock_t` and `uv_file`, with `uv_os_fd_t`.
+On Unix systems, this would be a `int fd` while on Windows it would be a `void* HANDLE`.
+
+## Details
+
+Original issue: [https://github.com/libuv/libuv/issues/856]()
+
+On Windows, a file descriptor is emulated in user-space as a wrapper around the real Win32 API.
+Indeed, the first operation libuv almost always must do is to throw away the wrapper (by calling `uv__get_osfhandle`).
+However, since the wrapper is implemented in user-space, this adds unnecessary overhead,
+makes the API less flexible, and is potentially error-prone
+(it assumes that the caller and libuv are linking against the same copy of msvcrt).
+A couple of functions already accept a `uv_os_sock_t`, essentially just to work around these limitations.
+While libuv could continuing adding dual APIs for everything (e.g. see `uv_poll_init` vs. `uv_poll_init_socket`),
```

Comment:
I assumed this is partly why libuv has separate APIs for handing sockets and other types of handles or files. This proposal wouldn't change any of that.